### ISSUE TITLE
Crash while saving tree cache (v0.3.0 bug)

### ISSUE THREAD

**Issue by ichiriac** [HvyIndustries/crane#218]
Hi,

After indexing Magento2 (18k Files) I have following errors :
```
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/home/devbox/Crane/projects/04a563704e43b8de181aac9f7588f62a/tree.tmp'
    at Error (native)
[Info  - 11:53:56] Connection to server got closed. Server will restart.
[Error - 11:54:01] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603 
```

To reproduce this, index Magento2, open a php file like autoload, and then type `$foo = new Magento\Admin ...`

Not sure, but the fact to debug into console each file parsed may decrease the total speed.

Seems also extracted classes are not registered under their namespace.

**Comment by nevadascout** [comment]
I'm looking into this now.

>Not sure, but the fact to debug into console each file parsed may decrease the total speed.

This should only happen if you have the `crane.debugMode` setting enabled

**Comment by nevadascout** [comment]
Hmm, this is working for me on Windows 10

Is this because (as the error says) `/home/devbox/Crane/projects/04a563704e43b8de181aac9f7588f62a/tree.tmp` doesn't exist?

**Comment by nevadascout** [comment]
This seems to be caused by the cache being corrupted during save - maybe because the project contains over 16,000 classes?

**Comment by ichiriac** [comment]
Found same kind of crash on a small project, without refreshing sources.

Here the log : 

```
fs.js:1099
  return binding.unlink(pathModule._makeLong(path));
                 ^

Error: ENOENT: no such file or directory, unlink '/home/devbox/Crane/projects/55fd9abfd0e3217dcf579e16246db47c/tree.tmp'
    at Error (native)
    at Object.fs.unlinkSync (fs.js:1099:18)
    at WriteStream.<anonymous> (/home/devbox/.vscode/extensions/HvyIndustries.crane-0.3.0/server/server.js:316:28)
    at emitNone (events.js:91:20)
    at WriteStream.emit (events.js:185:7)
    at fs.js:2027:14
    at FSReqWrap.oncomplete (fs.js:123:15)
[Info  - 18:49:42] Connection to server got closed. Server will restart.
[Error - 18:49:43] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603 
[Error - 18:49:43] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603 
```

After I've requested an indexing and it found 57k files as it had also scanned my node_modules folders (relating with symbolic links to all php frameworks) - the parsing passes

**Comment by ichiriac** [comment]
I can reproduce this only on big projects, seems that depends on time to process.

I guess something like, the saveProjectTree is called twice :

* Call 1 saveProjectTree 
* Call 2 saveProjectTree 
* Start GZIP from call 1
* Start GZIP from call 2
* Unlink TMP from call 1
* Error TMP does not exists from call 2

A better way of handling this kind of event could be to use a timeout timer. If the second call is done before the timeout, it resets the timer (this change could avoid 2 consecutive calls, only the last would be processed). 

The timer interval MUST be greater than the writing process in order to avoid 2 process in the same time :

* Call 1 -> Start timer 1
* Call 2 (3 sec after) -> disable timer 1 -> start Timer 2
* Timer 2 (10 sec after) -> start flushing
* Call 3 -> start timer 3 
* Flushing max 10 sec

If the flush takes more than 10 sec, the call 3 will be triggered and could make some errors. A flag with the processing state could also improve the algorithm, before restarting flush of call 3, check the state, if processing, start a new timer 4 and wait another 10 sec.

**Comment by ichiriac** [comment]
I will implement it on https://github.com/glayzzle/crane/tree/bugfix/saveProjectTree and when it's ready I will send you a pull request

**Comment by ichiriac** [comment]
@nevadascout It's fixed but don't know where to send the pull request - maybe you need to create a 0.3.1 branch ?

**Comment by nevadascout** [comment]
Create the PR onto the `feature/better-namespace-support` branch

**Comment by nevadascout** [comment]
Can you confirm this is fixed now that #231 is merged?

**Comment by nevadascout** [comment]
This is now fixed in v0.3.1
